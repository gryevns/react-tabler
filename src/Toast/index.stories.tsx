import React from "react";

import Toast from "../Toast";

export default {
    title: "Toast"
};

const Basic = ({ show = true, children, ...props }) => (
    <Toast {...props}>
        <Toast.Header>
            <Toast.Title>Toast Title</Toast.Title>
            <small className="ml-3">11 mins ago</small>
        </Toast.Header>
        <Toast.Body>{children}</Toast.Body>
    </Toast>
);

export const Stacked = () => (
    <div style={{ position: "absolute", top: 0, right: 0 }}>
        <Basic delay={1000}>A</Basic>
        <Basic delay={2000}>B</Basic>
        <Basic>C</Basic>
    </div>
);

export const Timeout = () => (
    <Basic delay={1000} fade>
        Toast content.
    </Basic>
);
