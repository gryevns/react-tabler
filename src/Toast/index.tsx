import React from "react";
import cn from "classnames";

import Buttons from "../Buttons";

interface ContextState {
    onHide: () => void;
}

const ToastContext = React.createContext({} as ContextState);

export interface ToastTitleProps {
    className?: string;
    children: React.ReactNode;
}

const ToastTitle = ({ className, children }: ToastTitleProps) => {
    const classes = cn(className, "mr-auto");
    return <strong className={classes}>{children}</strong>;
};

export interface ToastHeaderProps {
    children?: React.ReactNode;
}

const ToastHeader = ({ children }: ToastHeaderProps) => {
    const { onHide } = React.useContext(ToastContext);
    return (
        <div className="toast-header">
            {children}
            <Buttons.Button
                className="close ml-2 mb-1"
                data-dismiss="toast"
                onClick={() => onHide()}
            >
                <span aria-hidden="true">&times;</span>
                <span className="sr-only">Close</span>
            </Buttons.Button>
        </div>
    );
};

export interface ToastBodyProps {
    children?: React.ReactNode;
}

const ToastBody = ({ children }: ToastBodyProps) => (
    <div className="toast-body">{children}</div>
);

export interface ToastProps {
    children: React.ReactNode;
    fade?: boolean;
    delay?: number;
}

const Toast = ({ children, fade, delay }: ToastProps) => {
    const [show, setShow] = React.useState(true);
    const classes = cn("toast", { fade, show });
    if (delay) {
        React.useEffect(() => {
            const hideDelay = setTimeout(() => setShow(false), delay);
            return () => {
                clearTimeout(hideDelay);
            };
        }, []);
    }
    if (!show) {
        return null;
    }
    return (
        <ToastContext.Provider value={{ onHide: () => setShow(false) }}>
            <div
                className={classes}
                role="alert"
                aria-live="assertive"
                aria-atomic="true"
            >
                {children}
            </div>
        </ToastContext.Provider>
    );
};

Toast.Header = ToastHeader;
Toast.Title = ToastTitle;
Toast.Body = ToastBody;

export default Toast;
