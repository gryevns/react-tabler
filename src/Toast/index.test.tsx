import React from "react";
import { shallow, mount } from "enzyme";

import Toast from "./index";

describe("Toast", () => {
    it("should match toast snapshot", () => {
        const wrapper = shallow(
            <Toast>
                <Toast.Header>
                    <Toast.Title>title</Toast.Title>
                </Toast.Header>
                <Toast.Body>body</Toast.Body>
            </Toast>
        );
        expect(wrapper).toMatchSnapshot();
    });
    it("should match header snapshot", () => {
        const wrapper = shallow(
            <Toast.Header>
                <Toast.Title>title</Toast.Title>
            </Toast.Header>
        );
        expect(wrapper).toMatchSnapshot();
    });
    it("should match title snapshot", () => {
        const wrapper = shallow(<Toast.Title>title</Toast.Title>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match body snapshot", () => {
        const wrapper = shallow(<Toast.Body>body</Toast.Body>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should hide on close", () => {
        const wrapper = mount(
            <Toast>
                <Toast.Header>
                    <Toast.Title>title</Toast.Title>
                </Toast.Header>
                <Toast.Body>body</Toast.Body>
            </Toast>
        );
        wrapper.find("button").simulate("click");
        expect(wrapper.isEmptyRender()).toBe(true);
    });
});
