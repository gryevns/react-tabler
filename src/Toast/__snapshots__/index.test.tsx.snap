// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should match body snapshot 1`] = `
<div
  className="toast-body"
>
  body
</div>
`;

exports[`Toast should match header snapshot 1`] = `
<div
  className="toast-header"
>
  <ToastTitle>
    title
  </ToastTitle>
  <ForwardRef
    className="close ml-2 mb-1"
    data-dismiss="toast"
    onClick={[Function]}
  >
    <span
      aria-hidden="true"
    >
      Ã—
    </span>
    <span
      className="sr-only"
    >
      Close
    </span>
  </ForwardRef>
</div>
`;

exports[`Toast should match title snapshot 1`] = `
<strong
  className="mr-auto"
>
  title
</strong>
`;

exports[`Toast should match toast snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "onHide": [Function],
    }
  }
>
  <div
    aria-atomic="true"
    aria-live="assertive"
    className="toast show"
    role="alert"
  >
    <ToastHeader>
      <ToastTitle>
        title
      </ToastTitle>
    </ToastHeader>
    <ToastBody>
      body
    </ToastBody>
  </div>
</ContextProvider>
`;
